### PR TITLE
Feature/offshore development geomorphology

### PR DESCRIPTION
**Pull request by cfrontin**
Added a generic geomorphology class to handle both terrain and bathymetry data, for broad compatibility with gradients, etc. Begins work towards #62 and continues progress on #40.

### CODE REVIEW

File: ard/geomorphology.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    """
+    A class to represent geomorphology data for a given wind farm site domain.
+
+    Represents either bathymetry data for offshore sites or topography data for
+    onshore sites.
+    """
+
+    x_data = np.atleast_1d([0.0])  # x location in km
+    y_data = np.atleast_1d([0.0])  # y location in km
+    depth_data = np.atleast_1d([0.0])  # depth in m
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply ard/geomorphology.py:19]: Also, much of the work in this could be generalized to be possibly used elsewhere (3D interpolation and placed in utils)

**Review by jaredthomas68 — CHANGES_REQUESTED** [review]
Nice work Cory, this is coming along nicely. There are a few things that warrant more discussion before we move forward I think.

File: ard/__init__.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 from . import layout
 from . import offshore
 
+from . import geomorphology
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/__init__.py:5]: I was actually thinking about these sorts of things too- perhaps geographic constraints (`geography`?) is the right way to describe them? I did land on the conclusion (reflected here) that geomorphology was in some way distinct, but i like the idea of GIS layers being the right way to think about it with geomorphology being the fundamental layer. is GIS the right term? i suppose we do have in some sense a geographic information system. `geographic`?

File: ard/geomorphology.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        """
+        Get the shape of the geomorphology data.
+
+        Returns
+        -------
+        tuple
+            The shape of the geomorphology data.
+        """
+
+        self.check_valid()  # ensure that the current data is valid
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/geomorphology.py:53]: basically i wanted to make sure the grids all matched in their shape before returning x shape alone. might be overkill, i suppose

File: ard/geomorphology.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+            self.material_data = material_data_in.copy()
+
+        self.check_valid()  # ensure that the input data is valid
+
+    def get_depth_data(self):
+        """Get the depth at a given location."""
+        return self.depth_data
+
+
+class GeomorphologyGridData(GeomorphologyData):
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/geomorphology.py:92]: this was my overkill moment: i saw an opportunity to do uncertain and unstructured sounding data in the future. but Pietro seemed to think that the bathymetry and terrain data are almost always gonna be pretty crystal clear and not really a source of uncertainty? if true then yeah it should just be grids. are you aware if that's correct?

File: ard/geomorphology.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+            The depth at the given locations
+        """
+
+        x_query = np.atleast_1d(x_query)  # ensure x_query is a 1D array
+        y_query = np.atleast_1d(y_query)  # ensure y_query is a 1D array
+
+        if interp_method == "gaussian_process":
+            if self._gpr_eval is None:
+                from sklearn.gaussian_process import GaussianProcessRegressor
+                from sklearn.gaussian_process.kernels import Matern
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/geomorphology.py:141]: i will change it though, i think. hopefully shouldn't be that heavyweight to include skl.

File: ard/geomorphology.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+        # save into the geomorphology data object
+        self.y_data, self.x_data = np.meshgrid(y_coord, x_coord)
+        self.depth_data = grid_bathy
+
+        self.check_valid()  # make sure the loaded file is legit before exiting
+
+
+class TopographyGridData(GeomorphologyGridData):
+    """
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/geomorphology.py:286]: i feel you re:inheritence. the CFD code I worked with in grad school was crafted by a man who got addicted to C++ templates... similar pain point. the one reason i set it up this way for the purposes of clarity for end users, because it's not necessarily typical to use terrain and bathymetry data interchangably, and loaders are typically distinct between the two because those datasets are typically separated out. let me ask around casually to see if this is worth doing.

File: ard/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply ard/__init__.py:5]: As I've been working n the boundary constraint, I've realized that most of these constraints can be handled by the ray-casting algorithm once we have a 2-d projection defined by a set of ordered vertices. In this sense, topology and bathymetry are distinctly different from most of the other constraints, though we may put exclusion zones around really deep holes, etc. I think I'm fine with keeping the geomorphology here, but I think integrating a full GIS tool at the top level and feeding data from that into constraint definition would be a good goal for down the road.

[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/__init__.py:5]: I've flipped it over to `geographic` with the thinking that `geographic.py` can become a directory `geographic` in a non-breaking manner and then we can incorporate `geographic/geomorphology.py` and so on and so forth

File: ard/geomorphology.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/geomorphology.py:92]: collapsing then, rip fun

File: ard/geomorphology.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            )
+
+            return z_query  # return the depth at the given locations
+
+        elif interp_method == "spline":
+            # or, smooth bivariate spline from scipy implementation
+            from scipy.interpolate import SmoothBivariateSpline
+
+            # create the interpolator object
+            interpolator_sbs = SmoothBivariateSpline(
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply ard/geomorphology.py]: done!

File: test/unit/ard/test_geomorphology.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                    ),
+                )
+
+            # reset to a legitimate setup
+            self.geomorphology.set_values(
+                x_data_in=x_data,
+                y_data_in=y_data,
+                depth_data_in=depth_data,
+                material_data_in=material_data,
+            )
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply test/unit/ard/test_geomorphology.py]: i dont really understand? isnt the test gonna fail if that error is raised??

[the remark this answers is not in the document]
Comment (jaredthomas68) [reply test/unit/ard/test_geomorphology.py]: I guess this really is not needed, as long as there is a test the runs the code where the error would be raised if there were a problem

File: test/unit/ard/test_geomorphology.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        y_data, x_data = np.meshgrid([-1.0, 0.0, 1.0], [0.0, 2.0])
+        depth_data = np.ones_like(x_data)
+        material_data = np.array(
+            [["soil", "rock"], ["rock", "soil"], ["rock", "soil"]]
+        ).T
+
+        # set up a geomorphology grid data object
+        self.geomorphology = ard.geomorphology.GeomorphologyGridData()
+
+        for idx_case in range(4):
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply test/unit/ard/test_geomorphology.py]: done!

File: test/unit/ard/test_geomorphology.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            [-0.75, -0.5, -0.25, 0.25, 0.5, 0.75], [0.5, 1.5]
+        )
+        depth_sample = self.geomorphology.evaluate_depth(
+            x_sample.flatten(), y_sample.flatten()
+        )
+        # check that the values match a pyrite file
+        validation_data = {
+            "depth_sample": depth_sample,
+        }
+        ard.test_utils.pyrite_validator(
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply test/unit/ard/test_geomorphology.py]: so my philosophy with this is: the pyrite test should indicate "something changed" (beyond than machine precision jiggling, etc.) and gold-standard tests should indicate when you have something that's not matching an analytical expectation.

so, here, I'm inclined to leave it as it is (as in the updated code) but i think your point raises two "should do"s:
- make the pyrite test utility more clear when it fails
- write actual gold-standard tests for a lot of this

File: test/unit/ard/geographic/test_geomorphology.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+
+        # make sure the values are set in correctly
+        with subtests.test(f"set_values data equivalence tests"):
+            assert np.allclose(self.geomorphology.x_data, x_data)
+            assert np.allclose(self.geomorphology.y_data, y_data)
+            assert np.allclose(self.geomorphology.z_data, z_data)
+            assert np.allclose(self.geomorphology.get_z_data(), z_data)
+            assert np.all(self.geomorphology.get_shape() == x_data.shape)
+            assert self.geomorphology.material_data.size == 1
+            assert self.geomorphology.material_data == "soil"  # default value
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply test/unit/ard/geographic/test_geomorphology.py:101]: what if we used an `[0]` accessor since that should always be a numpy array

File: test/unit/ard/geographic/test_geomorphology.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        # set the values
+        self.geomorphology.set_values(
+            x_data_in=x_data,
+            y_data_in=y_data,
+            z_data_in=z_data,
+        )
+
+        # make sure the values are set in correctly
+        with subtests.test(f"set_values data equivalence tests"):
+            assert np.allclose(self.geomorphology.x_data, x_data)
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:95]: These assertions should be broken up into their own subtests so each subtest has only a single assertion

File: test/unit/ard/geographic/test_geomorphology.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply test/unit/ard/geographic/test_geomorphology.py]: I think that would be fine

File: test/unit/ard/geographic/test_geomorphology.py [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+            assert self.geomorphology.material_data.size == 1
+            assert self.geomorphology.material_data == "soil"  # default value
+
+        with subtests.test(f"check_valid final test"):
+            assert self.geomorphology.check_valid()  # check if the data is valid
+
+    def test_set_values_material(self, subtests):
+
+        # create a mesh and try to upload it
+        y_data, x_data = np.meshgrid([-1.0, 0.0, 1.0], [0.0, 2.0])
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:111]: x and y are swapped. I get that it works this way, but it is an odd way to write it.

File: test/unit/ard/geographic/test_geomorphology.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+        self.geomorphology.set_values(
+            x_data_in=x_data,
+            y_data_in=y_data,
+            z_data_in=z_data,
+            material_data_in=material_data,
+        )
+
+        # make sure the values are set in correctly
+        with subtests.test(f"set_values data equivalence tests"):
+            assert np.allclose(self.geomorphology.x_data, x_data)
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:129]: need to break up so each assert is its own subtest

File: test/unit/ard/geographic/test_geomorphology.py [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+            assert np.all(self.geomorphology.get_shape() == x_data.shape)
+
+        with subtests.test(f"check_valid final test"):
+            assert self.geomorphology.check_valid()  # check if the data is valid
+
+    def test_evaluate_default(self, subtests):
+
+        # create a mesh and try to upload it
+        y_data, x_data = np.meshgrid(
+            np.linspace(-1.0, 1.0, 5), np.linspace(0.0, 2.0, 5)
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:143]: backwards again

File: test/unit/ard/geographic/test_geomorphology.py [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+        # set the values
+        self.geomorphology.set_values(
+            x_data_in=x_data,
+            y_data_in=y_data,
+            z_data_in=z_data,
+        )
+
+        # grab the depth at points in the mesh domain
+        y_sample, x_sample = np.meshgrid(
+            [-0.75, -0.5, -0.25, 0.25, 0.5, 0.75], [0.5, 1.5]
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:159]: backwards x and y

File: test/unit/ard/geographic/test_geomorphology.py [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+            depth_sample_2 = self.geomorphology.evaluate(
+                x_sample.flatten(), y_sample.flatten()
+            )
+            assert np.allclose(depth_sample, depth_sample_2)
+            assert id_initial == id(self.geomorphology._interpolator_device)
+
+    def test_evaluate_gaussian(self, subtests):
+
+        # create a mesh and try to upload it
+        y_data, x_data = np.meshgrid([-1.0, 0.0, 1.0], [0.0, 2.0])
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:205]: reverse, reverse

File: test/unit/ard/geographic/test_geomorphology.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+
+        # make sure the data matches the statistical properties of the original data
+        with subtests.test(f"moorpy load statistics test"):
+            validation_data = {
+                "min": np.min(self.bathymetry.z_data),
+                "max": np.max(self.bathymetry.z_data),
+                "mean": np.mean(self.bathymetry.z_data),
+                "std": np.std(self.bathymetry.z_data),
+            }
+            ard.utils.test_utils.pyrite_validator(
```
Comment (jaredthomas68) [inline test/unit/ard/geographic/test_geomorphology.py:304]: This would be an easy test to write values out for, which makes it much easier to debug the test. I can role with it anyway, but saving test values in a non-human-readable format rubs me. the wrong way, especially when it is just a few values.

File: test/unit/ard/geographic/test_geomorphology.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+
+            # make sure check valid raises an exception
+            with subtests.test(f"check_valid bad override assertion test {idx_case}"):
+                with pytest.raises(AssertionError):
+                    assert self.geomorphology.check_valid()
+
+    def test_set_values(self, subtests):
+
+        # create a mesh and try to upload it
+        y_data, x_data = np.meshgrid([-1.0, 0.0, 1.0], [0.0, 2.0])
```
[the remark this answers is not in the document]
Comment (cfrontin) [reply test/unit/ard/geographic/test_geomorphology.py:81]: this is to get the right row/column ordering so you don't have to do a transpose to get it in the right orientation. this is really just to match the paradigm in the way the bathymetry files load with the benefit of also matching the major orientation of pyplot.

File: test/unit/ard/geographic/test_geomorphology.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (cfrontin) [reply test/unit/ard/geographic/test_geomorphology.py]: i was trying to see if copilot was gonna answer that lol.

File: test/unit/ard/geographic/test_geomorphology.py (hunk 18 above)
Comment (cfrontin) [reply test/unit/ard/geographic/test_geomorphology.py:304]: ok yeah you're definitely right. i think i mis-registered which test this comment was on due to some line numbers sliding around but on the statistics one it makes way more sense.
